In [220]:
import numpy as np
import pandas as pd

In [221]:
df = pd.read_csv("heart.csv")
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


In [222]:
df.shape

(1025, 14)

In [223]:
df.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,1025.000000,1025.000000,1025.000000,1025.000000,1025.00000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000,1025.000000
mean,54.434146,0.695610,0.942439,131.611707,246.00000,0.149268,0.529756,149.114146,0.336585,1.071512,1.385366,0.754146,2.323902,0.513171
std,9.072290,0.460373,1.029641,17.516718,51.59251,0.356527,0.527878,23.005724,0.472772,1.175053,0.617755,1.030798,0.620660,0.500070
min,29.000000,0.000000,0.000000,94.000000,126.00000,0.000000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,48.000000,0.000000,0.000000,120.000000,211.00000,0.000000,0.000000,132.000000,0.000000,0.000000,1.000000,0.000000,2.000000,0.000000
50%,56.000000,1.000000,1.000000,130.000000,240.00000,0.000000,1.000000,152.000000,0.000000,0.800000,1.000000,0.000000,2.000000,1.000000
75%,61.000000,1.000000,2.000000,140.000000,275.00000,0.000000,1.000000,166.000000,1.000000,1.800000,2.000000,1.000000,3.000000,1.000000
max,77.000000,1.000000,3.000000,200.000000,564.00000,1.000000,2.000000,202.000000,1.000000,6.200000,2.000000,4.000000,3.000000,1.000000


In [224]:
df.describe().T[['min', 'max', 'mean', 'std']]

,min,max,mean,std
age,29.0,77.0,54.434146,9.072290
sex,0.0,1.0,0.695610,0.460373
cp,0.0,3.0,0.942439,1.029641
trestbps,94.0,200.0,131.611707,17.516718
chol,126.0,564.0,246.000000,51.592510
fbs,0.0,1.0,0.149268,0.356527
restecg,0.0,2.0,0.529756,0.527878
thalach,71.0,202.0,149.114146,23.005724
exang,0.0,1.0,0.336585,0.472772
oldpeak,0.0,6.2,1.071512,1.175053


age (29-77), chol (126-564), oldpeak (0-6.2) gibi sütunların birbirinden çok farklı ölçeklerde olduğunu gördük. Bundan dolayı scale yapmamız lazım.

In [225]:
numeric_columns = df.select_dtypes(include='number').columns
numeric_columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='str')

lçeklenmesi gereken sütunları """age, trestbps, chol, thalach, oldpeak"""

In [226]:
numeric_cols = ['age', 'trestbps', 'chol', 'thalach', 'oldpeak']

In [227]:
df[numeric_cols].describe().T[['min', 'max', 'mean', 'std']]

,min,max,mean,std
age,29.0,77.0,54.434146,9.072290
trestbps,94.0,200.0,131.611707,17.516718
chol,126.0,564.0,246.000000,51.592510
thalach,71.0,202.0,149.114146,23.005724
oldpeak,0.0,6.2,1.071512,1.175053


In [228]:
from sklearn.preprocessing import MinMaxScaler

df_normalized = df.copy()
scaler_minmax = MinMaxScaler()
df_normalized[numeric_cols] = scaler_minmax.fit_transform(df[numeric_cols])

In [229]:
df_normalized[numeric_cols].describe().T[['min', 'max', 'mean', 'std']]

,min,max,mean,std
age,0.0,1.0,0.529878,0.189006
trestbps,0.0,1.0,0.354827,0.165252
chol,0.0,1.0,0.273973,0.117791
thalach,0.0,1.0,0.596291,0.175616
oldpeak,0.0,1.0,0.172825,0.189525


In [230]:
df = pd.read_csv("heart.csv")

In [231]:
from sklearn.preprocessing import StandardScaler

df_standardized = df.copy()
scaler_std = StandardScaler()
df_standardized[numeric_cols] = scaler_std.fit_transform(df[numeric_cols])

df_standardized[numeric_cols].describe().T[['min', 'max', 'mean', 'std']]

,min,max,mean,std
age,-2.804866,2.488552,-3.431402e-16,1.000488
trestbps,-2.148237,3.906079,-7.036106e-16,1.000488
chol,-2.327054,6.166694,-3.466062e-18,1.000488
thalach,-3.397080,2.299935,-4.332578e-16,1.000488
oldpeak,-0.912329,4.366603,-2.426243e-16,1.000488


Standardization'da her sütunun std'si eşitlenmiş (hepsi ~1), ama Normalization'da std'ler hâlâ birbirinden farklı (chol için 0.117, age için 0.189). Bu, önemli bir kavramsal fark: Normalization sadece min/max'ı sabitler, std'yi eşitlemez; Standardization ise std'yi eşitler, min/max'ı sabitlemez.

Bu tablo, bize gerçek hayatta şu soruyu sormamız gerektiğini gösteriyor: "Modelim için sütunların aynı ARALIKTA (0-1) olması mı önemli, yoksa aynı YAYILIMDA (std=1) olması mı önemli?" — cevap, kullandığınız algoritmaya göre değişir (Neural Network genelde 0-1 aralığını sever, Linear/Logistic Regression genelde std=1'i sever).

# 1. Train/Test Split ile İlişki — Data Leakage (Çok Kritik, Henüz Uygulamadık)

Şu ana kadar fit_transform()'u tüm veri setine uyguladık. Ama gerçek bir ML projesinde, önce veriyi train/test olarak ayırıp, scaler'ı sadece train'e fit etmeniz, sonra her ikisine de transform uygulamanız gerekir:

In [232]:
df = pd.read_csv("heart.csv")

In [233]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(df[numeric_cols], test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)  # SADECE train'e fit
X_test_scaled = scaler.transform(X_test)         # test'e sadece transform

# 2. inverse_transform() — Scale Edilmiş Veriyi Geri Çevirme

Bir model tahmin yaptıktan sonra, sonucu orijinal, anlaşılır ölçeğe geri döndürmeniz gerekebilir:

In [234]:
original_values = scaler.inverse_transform(df_standardized[numeric_cols])

# 3. Outlier'a Duyarlılık — Zaten Konuştuk Ama Burada Pratik Uygulaması Eksik

Hatırlarsanız Min-Max Scaling'in outlier'lara Standardization'dan bile daha hassas olduğunu konuşmuştuk. heart.csv'de chol sütununda (max=564, çoğu değer daha düşük) bir outlier olup olmadığına bakıp, RobustScaler ile karşılaştırma yapmadık:

In [235]:
from sklearn.preprocessing import RobustScaler

scaler_robust = RobustScaler()
df_robust = df.copy()
df_robust[numeric_cols] = scaler_robust.fit_transform(df[numeric_cols])

4. Modele Gerçek Etkisi — Henüz Test Etmedik

Şu ana kadar sadece sayıların nasıl değiştiğini gördük, ama "scaling gerçekten model performansını artırıyor mu" sorusuna henüz cevap vermedik. Örneğin target sütununu kullanarak bir Logistic Regression modelini hem scale edilmemiş hem scale edilmiş veriyle eğitip, accuracy/convergence hızı farkını görebiliriz — bu, öğrendiklerinizi somut bir sonuca bağlar.

5. Hangi Scaler'ı Ne Zaman Kullanacağınıza Dair Pratik Karar Süreci

Şu ana kadar teoride konuştuk ama heart.csv özelinde hangi scaler'ın bu modele/veri setine en uygun olduğuna hiç karar vermedik — bu, sadece "kod çalıştırmak" değil, "neden bunu seçtim" sorusuna cevap verebilmek demek, ki mülakatlarda tam olarak bu sorulur.

# İlk adım: Train/Test Split + Data Leakage'ı Önleme.

In [236]:
from sklearn.model_selection import train_test_split

X = df[numeric_cols]
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Train boyutu:", X_train.shape)
print("Test boyutu:", X_test.shape)

Train boyutu: (820, 5)
Test boyutu: (205, 5)


Neden y de ayırıyoruz? Çünkü az sonra bir model eğiteceğiz (Adım 2'de), model hem X (özellikler) hem y (hedef, yani target) ihtiyacı duyacak — bu yüzden ikisini birlikte, aynı satırlara karşılık gelecek şekilde bölmemiz gerekiyor.

# Adım 2 — Scaler'ı SADECE Train Verisine Fit Edelim

In [237]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

# fit_transform SADECE train'de
X_train_scaled = scaler.fit_transform(X_train)

# test'te SADECE transform (fit YOK!)
X_test_scaled = scaler.transform(X_test)

# Neden Bu Sıralama Kritik? (Hatırlatma)

Eğer yanlışlıkla şunu yapsaydık:

YANLIŞ KULLANIM - yapmayın

X_test_scaled = scaler.fit_transform(X_test)  # test'e de fit ediyor!

Bu, test verisinin kendi mean/std'sini kullanarak scale edilmesi demek olurdu — yani model, gerçek hayatta hiç görmeyeceği test verisinin istatistiksel özelliklerini (mean, std) önceden görmüş olurdu. Bu data leakage'dır ve modelin gerçek performansını olduğundan iyi göstermenize yol açar.

"Scaler'ın öğrendiği (fit ettiği) her şey, sadece TRAIN verisinden gelmeli — test verisi, sanki gerçek hayatta yeni gelen, daha önce hiç görülmemiş bir veriymiş gibi davranmalı."

Kontrol Edelim — Train ve Test'in Mean/Std'si Aynı mı?

In [238]:
import numpy as np

print("Train mean:", X_train_scaled.mean(axis=0).round(3))
print("Test mean:", X_test_scaled.mean(axis=0).round(3))

Train mean: [-0.  0. -0.  0. -0.]
Test mean: [ 0.037 -0.032  0.096 -0.052  0.137]


Beklenen sonuç: X_train_scaled'ın mean'i tam olarak 0'a yakın çıkacak (çünkü scaler ona göre fit edildi), ama X_test_scaled'ın mean'i 0'a yakın ama tam 0 değil olacak — bu normal ve doğru bir davranış, çünkü test verisi train'in istatistiklerine göre dönüştürüldü, kendi istatistiklerine göre değil.

Train Mean: [-0. 0. -0. 0. -0.] — Neden Sıfır?

Bu, matematiksel olarak kesin sıfır demek (yukarıda bahsettiğim gibi, -0. gösterimi, gerçek değerin -0.0000000001 gibi çok küçük bir sayı olmasından kaynaklanıyor, pratikte sıfır). Bu beklenen bir sonuç, çünkü Standardization'ın tanımı gereği, scaler'ın fit edildiği veri setinin mean'i her zaman tam olarak 0 çıkar — bu bir "başarı" değil, matematiksel bir zorunluluk, formülün doğası gereği böyle olmak zorunda.

Test Mean: [0.037, -0.032, 0.096, -0.052, 0.137] — Neden Sıfır DEĞİL?

İşte bu, bize çok önemli bir şey kanıtlıyor: Test verisinin mean'i 0'a yakın ama tam 0 değil. Bu, doğru ve beklenen bir sonuç, çünkü:

Scaler, train verisinin mean/std değerlerini öğrendi (fit etti)
Test verisine bu aynı, train'den öğrenilen dönüşümü uyguladı (transform)
Test verisinin kendi gerçek dağılımı, train verisininkinden hafifçe farklı olduğu için (rastgele bölündükleri için doğal bir varyasyon var), test verisinin mean'i tam sıfıra düşmedi
Bu Sonuç Neyi Kanıtlıyor?

Bu, data leakage'ın olmadığının kanıtı! Eğer yanlışlıkla scaler.fit_transform(X_test) yapsaydık, test mean'i de tam 0 çıkardı — çünkü scaler test'in kendi istatistiklerine göre fit edilmiş olurdu. Sizin sonucunuzda test mean'inin sıfırdan sapması, scaler'ın gerçekten sadece train'den öğrendiğini, test verisine "kör" bir şekilde (kendi istatistiklerine bakmadan) uygulandığını gösteriyor — yani doğru, sızıntısız bir pipeline kurmuşsunuz.

Hocam güzel, şimdi scaling'in gerçekten işe yarayıp yaramadığını somut olarak görelim.

Adım 3 — Scaling'siz vs Scaling'li Model Karşılaştırması

# 1. Önce Scaling YAPMADAN Bir Model Eğitelim

In [239]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import warnings
warnings.filterwarnings('ignore')  # convergence uyarılarını görmek için kapatmayalım aslında, açık bırakalım
#Aslında uyarıyı görmek istiyoruz, o yüzden warnings'i kapatmayın:

In [240]:
# SCALING YAPILMAMIŞ ham veriyle model
model_raw = LogisticRegression(max_iter=1000)
model_raw.fit(X_train, y_train)

y_pred_raw = model_raw.predict(X_test)
acc_raw = accuracy_score(y_test, y_pred_raw)

print("Scaling YOK - Accuracy:", acc_raw)
print("Scaling YOK - Kaç iterasyonda durdu:", model_raw.n_iter_)

Scaling YOK - Accuracy: 0.6585365853658537
Scaling YOK - Kaç iterasyonda durdu: [45]


# 2. Şimdi Scaling YAPILMIŞ Veriyle Aynı Modeli Eğitelim

In [241]:
# SCALING YAPILMIŞ veriyle model (bir önceki adımda hazırladığımız X_train_scaled/X_test_scaled)
model_scaled = LogisticRegression(max_iter=1000)
model_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = model_scaled.predict(X_test_scaled)
acc_scaled = accuracy_score(y_test, y_pred_scaled)

print("Scaling VAR - Accuracy:", acc_scaled)
print("Scaling VAR - Kaç iterasyonda durdu:", model_scaled.n_iter_)

Scaling VAR - Accuracy: 0.6585365853658537
Scaling VAR - Kaç iterasyonda durdu: [7]


Not: Scaling VAR sonucunda n_iter_=7 çıkması gerçekten çok hızlı bir yakınsama — bu iyi bir işaret.

Neyi Karşılaştıracağız?

1. Accuracy (doğruluk) — İki modelin tahmin başarısı birbirine yakın çıkabilir (çünkü heart.csv'deki sayısal aralıklar aşırı uçlarda değil), ama...

2. n_iter_ (kaç iterasyonda yakınsadığı) — İşte asıl fark burada ortaya çıkacak. Hatırlarsanız Standardization konusunda "gradient descent'in zikzak çizmeden, hızlı yakınsaması" konusunu konuşmuştuk — scaling YAPILMAMIŞ veride, modelin daha fazla iterasyona ihtiyaç duyduğunu (ya da hiç yakınsamadığını, bir uyarı verdiğini) görebiliriz.

1. Accuracy — Neden Aynı Çıktı?

İkisi de birebir aynı (0.6585) — bu şaşırtıcı gelebilir ama aslında beklenen bir durum olabilir. Sebebi: heart.csv'deki sayısal aralıklar (age 29-77, chol 126-564 gibi) aşırı uçlarda değil, yani modelin karar sınırını (decision boundary) bulma yeteneği üzerinde büyük bir çarpıklık yaratacak kadar dramatik bir ölçek farkı yok. Model, biraz daha "zorlanarak" da olsa, sonunda aynı doğru sonuca ulaşabiliyor.

2. n_iter_ — İşte Asıl Kanıt Burada!

Scaling YOK: 45 iterasyon vs Scaling VAR: 7 iterasyon — bu 6.4 kat bir fark! Bu, tam olarak daha önce teorik olarak konuştuğumuz şeyin somut kanıtı:

"Farklı ölçeklerdeki özellikler, gradient descent'in 'vadisini' eliptik/çarpık yapar, bu da modelin zikzak çizerek, daha yavaş yakınsamasına sebep olur."

Scaling YOK modelinin 45 iterasyona ihtiyaç duyması, tam olarak bu "zikzak çizme" sürecinin sayısal göstergesi — model, minimum noktaya ulaşmak için çok daha fazla adım atmak zorunda kaldı.

Bu Neden Önemli? (Pratik Sonuç)

Bu küçük veri setinde (1025 satır) fark hissedilmeyebilir — 45 iterasyon bile göz açıp kapayana kadar tamamlanır, gerçek bir "yavaşlık" hissetmezsiniz. Ama düşünün:

Milyonlarca satırlık bir veri setinde
Çok daha fazla özellik (feature) içeren bir modelde
Neural Network gibi, her iterasyonun çok daha maliyetli olduğu bir modelde

Bu 6.4 kat fark, saatler süren bir eğitim sürecini dakikalara indirmek anlamına gelebilir — yani scaling'in faydası, küçük ve basit veri setlerinde "gizli" kalıyor ama büyük, karmaşık sistemlerde çok ciddi bir zaman/maliyet tasarrufu sağlıyor.

Özetle Ne Öğrendik?
Accuracy her zaman değişmeyebilir — scaling, modelin "doğru cevabı bulup bulamayacağını" değil, "o cevaba ne kadar hızlı ulaştığını" etkiler (özellikle basit/küçük veri setlerinde)
n_iter_ farkı, scaling'in gerçek faydasının somut kanıtıdır — teoride konuştuğumuz "zikzak çizme, yavaş yakınsama" sorununu, gerçek bir sayıyla (45 vs 7) doğrulamış olduk
Ağaç tabanlı modellerde (Decision Tree, Random Forest) bu fark hiç görülmezdi — çünkü onlar gradient descent kullanmaz, scaling'e ihtiyaç duymazlar (hatırlarsanız bunu da konuşmuştuk)

# RobustScaler karşılaştırması.

Önce Outlier Var mı Kontrol Edelim

RobustScaler'ın gücünü görmek için, önce chol (kolesterol) sütununda gerçekten outlier olup olmadığına bakalım:

In [242]:
Q1 = df['chol'].quantile(0.25)
Q3 = df['chol'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5*IQR
upper = Q3 + 1.5*IQR

outliers = df[(df['chol'] < lower) | (df['chol'] > upper)]
print("Outlier sayısı:", len(outliers))
print(outliers['chol'].values)

Outlier sayısı: 16
[417 564 409 564 394 407 564 407 394 394 409 417 407 407 417 409]


chol sütununun max değeri 564 iken çoğu değer 200-300 civarında olduğu için, muhtemelen birkaç outlier çıkacak.

# RobustScaler Uygulayalım

In [243]:
from sklearn.preprocessing import RobustScaler

scaler_robust = RobustScaler()

X_train_robust = scaler_robust.fit_transform(X_train)
X_test_robust = scaler_robust.transform(X_test)

# Üç Yöntemi Yan Yana Karşılaştıralım

In [244]:
comparison = pd.DataFrame({
    'Standardization_mean': X_train_scaled.mean(axis=0),
    'Standardization_std': X_train_scaled.std(axis=0),
    'Robust_median': pd.DataFrame(X_train_robust, columns=numeric_cols).median().values,
    'Robust_std': X_train_robust.std(axis=0)
}, index=numeric_cols)

print(comparison)

          Standardization_mean  Standardization_std  Robust_median  Robust_std
age              -1.299773e-17                  1.0            0.0    0.654380
trestbps          2.989479e-16                  1.0            0.0    0.880385
chol             -1.819683e-16                  1.0            0.0    0.787906
thalach           2.946153e-16                  1.0            0.0    0.669013
oldpeak          -3.357748e-17                  1.0            0.0    0.691168


1. Robust_median = 0.0 — Bu Doğru mu?

Evet, tamamen beklenen bir sonuç.

Scaler'ı train verisine fit ettiğinizde, formül gereği medyanı kendisinden çıkarıyorsunuz (median - median = 0), bu yüzden train verisinin medyanı matematiksel olarak her zaman tam 0 çıkar — tıpkı Standardization'da train mean'inin her zaman 0 çıkması gibi, burada da aynı mantık medyan için geçerli.

Robust_std Sütunu — Asıl İlginç Kısım Burada

Dikkat edin, hiçbir sütunun std'si tam 1 değil:

age:      0.654
trestbps: 0.880
chol:     0.788
thalach:  0.669
oldpeak:  0.691

Standardization'da hepsi tam 1.000 çıkmıştı, ama RobustScaler'da hiçbiri 1'e eşit değil — neden?

Çünkü RobustScaler, std'yi 1'e eşitlemeyi hiç hedeflemez — o sadece medyan ve IQR'a göre ölçeklendirme yapar, std'nin ne çıkacağı verinin gerçek dağılımına bağlı, garanti edilmiş bir hedef değil. Bu, RobustScaler'ın Standardization'dan kavramsal olarak farklı bir amaç güttüğünü gösteriyor — biri "std=1" garantisi verirken, diğeri "outlier'lara dayanıklı, medyan bazlı" bir ölçeklendirme garantisi veriyor, ikisi farklı hedeflere optimize edilmiş.

# Ne Görmemizi Bekliyoruz?

Özellikle chol satırına dikkat edin — Standardization'da Standard_std tam 1.0'a yakın çıkacak, ama RobustScaler'da Robust_std 1'den belirgin şekilde farklı (muhtemelen daha büyük) çıkacak. Bunun sebebi:

Standardization, mean/std hesaplarken 16 outlier'ı da dahil eder, bu da std'yi şişirir (gerçek "tipik" yayılımdan daha büyük gösterir)
RobustScaler, medyan/IQR kullandığı için outlier'lardan etkilenmez, "tipik" veri noktalarının gerçek yayılımını daha saf şekilde yansıtır

# Model Performansını da Test Edelim

In [245]:
model_robust = LogisticRegression(max_iter=1000)
model_robust.fit(X_train_robust, y_train)

y_pred_robust = model_robust.predict(X_test_robust)
acc_robust = accuracy_score(y_test, y_pred_robust)

print("RobustScaler - Accuracy:", acc_robust)
print("RobustScaler - Kaç iterasyonda durdu:", model_robust.n_iter_)

RobustScaler - Accuracy: 0.6585365853658537
RobustScaler - Kaç iterasyonda durdu: [8]


3. Accuracy Yine Aynı — Neden?

Üç yöntemde de accuracy 0.6585 ile birebir aynı çıktı. Bu, heart.csv özelinde şunu gösteriyor: Logistic Regression'ın karar sınırını bulma yeteneği, bu veri setinde hangi scaling yöntemini kullandığınızdan bağımsız — 16 outlier olsa bile, bu outlier'lar modelin doğru sınıflandırma yapmasını engelleyecek kadar baskın/zararlı değil.

4. n_iter_ — En Merak Edilen Kısım
Scaling YOK:      45 iterasyon
Standardization:   7 iterasyon
RobustScaler:       8 iterasyon

Standardization ve RobustScaler neredeyse aynı hızda (7 vs 8) yakınsadı, ikisi de Scaling YOK'a göre çok daha hızlı. Bu bize şunu gösteriyor: gradient descent'in hızlanması için asıl önemli olan şey, "tüm sütunların BENZER bir ölçekte olması" — ister Standardization'ın std=1 mantığıyla, ister RobustScaler'ın medyan/IQR mantığıyla olsun, her ikisi de sütunları "karşılaştırılabilir" bir büyüklüğe getirdiği için, ikisi de yakınsama hızını benzer şekilde iyileştiriyor.

Peki RobustScaler'ı Ne Zaman Tercih Etmeliyiz? (Bu Örnekten Çıkardığımız Ders)

Bu deneyde accuracy ve n_iter_ açısından RobustScaler ile Standardization arasında pratik bir fark görmedik — çünkü heart.csv'deki 16 outlier, modelin genel başarısını bozacak kadar güçlü değildi. Ama eğer:

Outlier sayısı çok daha fazla olsaydı
Outlier'lar çok daha aşırı değerler taşısaydı (564 yerine, örneğin 5000 gibi)
Ya da mesafe tabanlı bir model (KNN, SVM gibi, Logistic Regression'dan farklı olarak) kullansaydık

O zaman RobustScaler ile Standardization arasındaki fark çok daha belirgin ortaya çıkardı. Bu deney bize önemli bir pratik ders veriyor: "outlier var" demek, otomatik olarak "RobustScaler kesin daha iyi sonuç verir" anlamına gelmiyor — etkisi, outlier'ların şiddetine, sayısına ve kullanılan modele bağlı.

Genel Özet — Bu Deney Dizisinden Çıkardığımız 3 Ana Ders:

Scaling, accuracy'yi her zaman değiştirmez, ama yakınsama hızını (n_iter_) dramatik şekilde etkileyebilir (45 → 7-8)
Standardization ve RobustScaler, farklı matematiksel hedeflere sahip (std=1 garantisi vs medyan/IQR bazlı dayanıklılık), ama pratikte benzer hız kazancı sağlayabilirler
"Hangi scaler daha iyi" sorusunun kesin bir cevabı yok — veri setinin outlier yoğunluğuna, modelin türüne ve gerçek performans testine bağlı, bu yüzden elle deneyip karşılaştırmak (tam da şimdi yaptığımız gibi) en güvenilir yöntem

# 1. Accuracy 0.65 Ne Anlama Geliyor? (Kavramsal Açıklama)

Accuracy, modelin test setindeki tüm tahminlerin yüzde kaçını doğru bildiğini gösteren bir toplu/genel başarı ölçüsüdür — tek bir yeni değer için "bu tahminin doğru olma ihtimali %65" anlamına gelmez.

Somut örnek: Test setinizde 205 kişi vardı (1025'in %20'si). accuracy=0.65 demek, bu 205 kişiden yaklaşık 133'ünün (0.65 × 205) doğru sınıflandırıldığı, 72'sinin ise yanlış sınıflandırıldığı anlamına gelir. Yani bu, geçmişte test ettiğiniz grubun genel başarı oranı, yeni bir bireysel tahmin için "bu kişiye özel %65 güven" demek değil.

In [246]:
yeni_kisi = [[55, 130, 250, 150, 1.5]]  # age, trestbps, chol, thalach, oldpeak
tahmin = model_scaled.predict(scaler.transform(yeni_kisi))
print(tahmin)  # 0 ya da 1 (kesin bir sınıf çıktısı, "yüzde" değil)

[0]


Model, bu kişi için kesin bir sınıf (0 = hastalık yok, 1 = hastalık var) döndürür — "%65 ihtimalle 1" demez

In [247]:
olasilik = model_scaled.predict_proba(scaler.transform(yeni_kisi))
print(olasilik)  # örneğin [[0.42, 0.58]] gibi, sınıf 0 ve sınıf 1 olasılıkları

[[0.58239934 0.41760066]]


Bu predict_proba() çıktısı, o spesifik kişi için modelin ne kadar "emin" olduğunu gösterir — ama bu da genel accuracy'den (0.65) farklı bir şey, her tahmin için kendi başına hesaplanır.

2. Asıl Dikkat Etmeniz Gereken Nokta — 0.65 Aslında Düşük Bir Sonuç

Hocam burada size dürüst bir şey söylemem lazım: heart.csv için 0.65 accuracy, aslında beklenenden düşük bir sonuç. Bu veri setiyle literatürde genelde %80-90 arası accuracy elde edilir. Sizin sonucunuzun düşük çıkmasının muhtemel sebebi şu:

Biz sadece 5 sayısal sütunu kullandık (age, trestbps, chol, thalach, oldpeak), ama heart.csv'deki cp (göğüs ağrısı tipi), thal, ca, exang gibi binary/kategorik sütunlar da aslında kalp hastalığı tahmininde çok güçlü belirleyiciler — biz bunları scaling pratiği yaparken bilinçli olarak dışarıda bırakmıştık (hatırlarsanız "bunlar zaten 0-1 arası, scale etmeye gerek yok" demiştik), ama bu, modelin onları hiç kullanmadığı anlamına geliyor, sadece 5 sütunla eğittik.

In [248]:
all_features = df.drop('target', axis=1)  # target hariç TÜM sütunlar
X_train_full, X_test_full, y_train_full, y_test_full = train_test_split(
    all_features, y, test_size=0.2
)

scaler_full = StandardScaler()
X_train_full_scaled = scaler_full.fit_transform(X_train_full)
X_test_full_scaled = scaler_full.transform(X_test_full)

model_full = LogisticRegression(max_iter=1000)
model_full.fit(X_train_full_scaled, y_train_full)

acc_full = accuracy_score(y_test_full, model_full.predict(X_test_full_scaled))
print("TÜM sütunlarla Accuracy:", acc_full)

TÜM sütunlarla Accuracy: 0.8439024390243902


Bu, %14'lük bir iyileşme — sadece kategorik/binary sütunları da modele dahil ederek bu kadar büyük bir fark elde ettik.

Bu Sonuç Bize Ne Öğretiyor?

1. Feature Selection'ın (Sütun Seçiminin) Önemi
Hatırlarsanız daha önce correlation analysis ve feature selection konusunu konuşmuştuk — işte tam olarak bunun canlı kanıtı: cp (göğüs ağrısı tipi), thal, ca, exang gibi sütunlar, kalp hastalığı tahmininde gerçekten güçlü belirleyiciler olduğu için, onları dışarıda bırakmak modelin gücünü ciddi şekilde kısıtlamış.

2. Scaling Tek Başına Yeterli Değil
Bu bize önemli bir ders veriyor: Scaling, modelin yakınsama HIZINI iyileştirir, ama modelin "doğru bilgiyi görüp görmediği" ayrı bir konu. Siz mükemmel scaling yapsanız bile, eğer modele önemli bilgiyi taşıyan sütunları vermezseniz, model iyi bir tahmin gücüne ulaşamaz. Scaling bir "optimizasyon" aracıdır, "bilgi tamamlama" aracı değildir.

3. Gerçek ML İş Akışının Nasıl İşlediğine Dair Küçük Bir Önizleme
Bu, aslında bize feature selection konusunun (hangi sütunları modele dahil edeceğinize karar verme süreci) neden bu kadar kritik olduğunu somut bir sayıyla gösterdi — %14'lük bir accuracy artışı, gerçek bir projede çok büyük bir fark demek.

# Logistic Regression'ın katsayılarına (coefficients) bakarak, hangi sütunun tahmine en çok etki ettiğini görebiliriz

In [249]:
coef_df = pd.DataFrame({
    'Feature': all_features.columns,
    'Coefficient': model_full.coef_[0]
}).sort_values('Coefficient', key=abs, ascending=False)

print(coef_df)

     Feature  Coefficient
2         cp     0.873620
1        sex    -0.838364
11        ca    -0.749551
9    oldpeak    -0.641676
7    thalach     0.574649
12      thal    -0.518392
8      exang    -0.457595
10     slope     0.364809
3   trestbps    -0.290966
4       chol    -0.277340
6    restecg     0.231138
5        fbs     0.016458
0        age    -0.008719


Hocam harika bir sonuç geldi, bunu detaylıca yorumlayalım.

## Sonuçları Büyükten Küçüğe Sıralı Görelim

| Sütun | Katsayı | Etki Yönü |
|---|---|---|
| **cp** | +0.874 | En güçlü, **pozitif** |
| **sex** | -0.838 | Çok güçlü, **negatif** |
| **ca** | -0.750 | Güçlü, **negatif** |
| **oldpeak** | -0.642 | Güçlü, **negatif** |
| **thalach** | +0.575 | Orta-güçlü, **pozitif** |
| **thal** | -0.518 | Orta-güçlü, **negatif** |
| **exang** | -0.458 | Orta, **negatif** |
| **slope** | +0.365 | Orta, **pozitif** |
| **trestbps** | -0.291 | Zayıf, **negatif** |
| **chol** | -0.277 | Zayıf, **negatif** |
| **restecg** | +0.231 | Zayıf, **pozitif** |
| **fbs** | +0.016 | Neredeyse **etkisiz** |
| **age** | -0.009 | Neredeyse **etkisiz** |

## En Kritik Gözlem — `age` ve `fbs` Neredeyse Sıfır!

Bu, gerçekten **şaşırtıcı ve öğretici** bir sonuç: **`age` (yaş)**, sezgisel olarak kalp hastalığında çok önemli görünse de, bu modelde katsayısı **-0.009**, yani **neredeyse hiç etkisi yok**! Aynı şekilde `fbs` (açlık kan şekeri) de **0.016** ile pratik olarak **anlamsız**.

**Neden böyle olabilir?** Muhtemelen `age`'in etkisi, **diğer sütunlarla (özellikle `thalach`, `oldpeak`, `ca` gibi) zaten dolaylı olarak temsil ediliyor** — yani yaşın kendisi değil, **yaşın yol açtığı fizyolojik sonuçlar** (maksimum kalp atış hızının düşmesi, damar tıkanıklığı sayısının artması gibi) modele **doğrudan** veriliyor, bu da `age`'in **kendi başına** ek bir bilgi katmasını gereksiz kılıyor. Bu, istatistikte **multicollinearity**'nin (değişkenler arası ilişki) bir sonucu olabilir — hatırlarsanız correlation analysis konusunda bahsetmiştik.

## En Güçlü İki Sütun — `cp` ve `sex`

**`cp` (göğüs ağrısı tipi):** **+0.874** ile en güçlü pozitif etki — bu **tıbbi olarak da çok mantıklı**, göğüs ağrısının tipi, kalp hastalığı teşhisinde klasik olarak en önemli belirteçlerden biridir.

**`sex` (cinsiyet):** **-0.838** ile ikinci en güçlü, ama **negatif**. Bu, veri setinde `sex` muhtemelen `1=erkek, 0=kadın` olarak kodlanmış, ve negatif katsayı **"erkek olmanın, bu modelde hastalık olasılığını azalttığı"** anlamına geliyor gibi görünüyor — bu **ilk bakışta tıbbi sezgilere ters** durabilir (genelde erkeklerde kalp hastalığı riski istatistiksel olarak daha yüksek bilinir). Bu, muhtemelen **veri setinin kendine özgü demografik dağılımından** (belki bu spesifik veri setinde, hastalığı olan kadınların sayısı orantısız şekilde fazla) kaynaklanıyor olabilir — bu tür "sezgiye ters" sonuçlar, gerçek dünyada **veri setinin toplama şeklini sorgulamanız** gerektiğinin bir işaretidir.

## Bu Sonuçtan Çıkardığımız Genel Ders — Feature Importance'ın Gücü

Bu analiz bize gösteriyor ki:
1. **Her sütun eşit değerde değil** — `cp`, `sex`, `ca` gibi sütunlar **modelin kararını domine ederken**, `age`, `fbs` gibi "sezgisel olarak önemli görünen" sütunlar aslında **çok az katkı** sağlıyor
2. Eğer bir gün **feature selection** yapmanız gerekirse (örneğin çok fazla sütunlu bir veri setinde, gereksiz olanları elemek için), bu tür bir **katsayı/importance analizi**, hangi sütunları **tutup hangilerini atacağınıza** karar vermenizde size **somut bir kılavuz** verir

## Küçük Bir Uyarı — Katsayı Büyüklüğü Yorumlanırken Dikkat

Bu katsayıları **doğrudan karşılaştırabiliyoruz** çünkü **Standardization uyguladık** — tüm sütunlar aynı ölçekte (std=1) olduğu için, katsayıların büyüklüğü **adil bir şekilde** karşılaştırılabilir. Eğer scaling **yapmasaydık**, büyük ölçekli bir sütun (örneğin `chol`, 126-564 arası) otomatik olarak **küçük bir katsayı** alırdı (çünkü zaten büyük sayılarla çarpılıyor), bu da **yanlış bir "önemsizlik" izlenimi** verirdi — işte bu, **scaling'in feature importance analizini de doğru şekilde yapabilmemiz için neden gerekli olduğunun** bir başka kanıtı.

---

**Özetle:** Bu tablo, hem `heart.csv`'nin **gerçek tıbbi dinamiklerine** bir pencere açtı (göğüs ağrısı tipi ve cinsiyetin baskın rolü gibi) hem de bize **scaling'in sadece hız için değil, aynı zamanda adil/yorumlanabilir bir feature importance analizi için de gerekli olduğunu** gösterdi. Bu, sizin şu ana kadarki tüm öğrenim sürecinizin (data cleaning → encoding → scaling → model → interpretation) **gerçek bir uçtan uca örneği** oldu, tebrikler!

Hocam peki biz dışarıdan yeni birey ekledik ve yüzde 70 oranla 0 sınıfında dedi. Biz bunun doğru olduğunu o anda nasıl öğrenebiliriz ?

Hocam çok temel ama gerçekten kritik bir soru — ve dürüst cevap şu: **o anda, doğrudan bilemezsiniz.**

## Neden Bilemeyiz?

Model, yeni bir kişi için **"%70 ihtimalle sınıf 0"** dediğinde, bu bir **tahmindir**, bir **kesinlik değil**. Bu kişinin **gerçek** durumu (gerçekten kalp hastası olup olmadığı) ancak **başka, bağımsız bir yöntemle** (örneğin gerçek bir tıbbi tetkik, anjiyo, doktor teşhisi) doğrulanabilir. Model, sadece **geçmiş verilerdeki örüntülere** bakarak bir **olasılık** sunuyor, **gerçek bir teşhis koymuyor**.

## Peki O Zaman Bu Tahmine Nasıl Güveniriz?

Burada devreye **iki farklı zaman çizelgesi** giriyor, ayırmamız gerekiyor:

### 1. Model Kurulurken — "Genel Güvenilirlik" (Bunu Zaten Yaptık)

Biz modeli **test setinde** (gerçek sonucu zaten bilinen 205 kişide) denedik ve **%79.5 accuracy** bulduk. Bu bize şunu söylüyor: *"Bu model, daha önce görmediği kişilerde ortalama olarak %79.5 doğru tahmin yapıyor."* Bu, modelin **genel güvenilirlik derecesini** ölçer — ama **belirli bir bireye özel** bir garanti değildir.

### 2. Yeni Birey Geldiğinde — "Anlık Doğrulama" (Bunun Cevabı: Yapılamaz)

Yeni, gerçek dünyadan gelen bir kişi için, **o anda** modelin doğru mu yanlış mı olduğunu bilmenin **hiçbir matematiksel yolu yoktur** — çünkü bu kişinin **gerçek etiketi (ground truth)** elinizde yok. Bu, ML'in **doğası gereği kabul etmeniz gereken bir sınırlamadır**.

## Peki Gerçek Hayatta Bu Nasıl Yönetiliyor?

**1. Modelin Güven Skorunu (Confidence) Dikkate Alma**
`%70 sınıf 0` demek, aslında `%30 sınıf 1` demek de oluyor — yani model **çok da emin değil**. Eğer model `%98 sınıf 0` deseydi, bu **çok daha güçlü bir sinyal** olurdu. Bu yüzden pratikte, **düşük güvenli tahminler** (örneğin %55-%70 arası, yani "yazı-tura"ya yakın) genelde **"belirsiz/gri bölge"** olarak işaretlenip, **ek insan incelemesi** gerektirir.

```python
olasilik = model_full.predict_proba(scaler_full.transform(yeni_kisi))
print(olasilik)  # [[0.70, 0.30]] gibi
```

**2. Modeli "Karar Verici" Değil, "Destekleyici Araç" Olarak Kullanma**
Gerçek tıbbi uygulamalarda, bu tür modeller **doktorun yerine geçmez**, doktora **"bu hastaya dikkat et, risk sinyali var"** diyen bir **ön tarama/destek aracı** olarak kullanılır. Nihai karar, **gerçek tetkiklerle** doğrulanır.

**3. Zamanla Geri Bildirim Toplama (Feedback Loop)**
Gerçek dünyada, bu kişi ileride **gerçek bir teşhis** aldığında (örneğin 6 ay sonra anjiyo yaptırdığında), o **gerçek sonuç** kayıt altına alınır ve modelin **o spesifik tahmini doğru muydu yanlış mıydı** geriye dönük olarak öğrenilir. Bu veriler biriktikçe, model **yeniden eğitilip** (retraining) güncellenebilir — bu, gerçek ML sistemlerinin **sürekli bir döngü** olarak çalışmasının sebebidir.

## Basit Bir Benzetmeyle Özetleyelim

Bunu bir **hava durumu tahmini** gibi düşünün: Meteorolog "yarın %70 ihtimalle yağmur yağacak" dediğinde, **o an** bu tahminin doğru olup olmadığını bilemezsiniz — sadece **yarın gerçekten** yağmur yağıp yağmadığına bakarak öğrenirsiniz. Ama meteoroloğun **geçmişteki** tahminlerinin **genel doğruluk oranına** (örneğin "bu meteorolog %80 doğru tahmin ediyor") bakarak, o kişinin **güvenilirliğine** dair bir fikir edinebilirsiniz — işte bizim `accuracy=0.7951` dediğimiz şey de tam olarak bu **"meteoroloğun genel güvenilirlik notu"**.

---

**Özetle:** Yeni bir bireysel tahminin **o anda** doğru olup olmadığını **asla kesin olarak bilemezsiniz** — sadece modelin **genel/geçmiş performansına** (accuracy, precision, recall gibi metriklere) bakarak, ona **ne kadar güvenmeniz gerektiğine** dair bir fikir edinebilirsiniz. Gerçek doğrulama, ancak **zaman geçtikten sonra, gerçek sonuç ortaya çıktığında** mümkün olur — bu, makine öğrenmesinin **kabul etmeniz gereken temel bir gerçeğidir**, "sihirli bir kesinlik makinesi" değildir.

In [294]:
import pandas as pd
from sklearn.model_selection import KFold

# Sample Data
data = {'City': ['London', 'Paris', 'London', 'Tokyo', 'Paris', 'London', 'Tokyo', 'Rome', 'Rome'],
        'Purchased': [1, 0, 0, 1, 1, 1, 0, 0, 1]}
df = pd.DataFrame(data)

In [295]:
df

,City,Purchased
0,London,1
1,Paris,0
2,London,0
3,Tokyo,1
4,Paris,1
5,London,1
6,Tokyo,0
7,Rome,0
8,Rome,1


In [296]:
df.shape

(9, 2)

In [297]:
target = 'Purchased'
feature = 'City'

In [ ]:
# Global mean for smoothing and potentially filling NAs later
global_mean = df[target].mean()
# Smoothing factor
m = 1

# Setup KFold
kf = KFold(n_splits=3, shuffle=True, random_state=42)
df[f'{feature}_encoded'] = 0.0 # Initialize encoded column

# Apply CV encoding
for train_index, val_index in kf.split(df):
    df_train, df_val = df.iloc[train_index], df.iloc[val_index]

    # Calculate smoothed means on the training part of the fold
    means = df_train.groupby(feature)[target].agg(['count', 'mean'])
    smoothed_means = (means['count'] * means['mean'] + m * global_mean) / (means['count'] + m)

    # Apply means to the validation part of the fold
    # Use .map() and fill potential new categories in validation with global mean
    df.loc[val_index, f'{feature}_encoded'] = df_val[feature].map(smoothed_means).fillna(global_mean)

print("Original DataFrame with CV Target Encoding:")
print(df)


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


In [309]:
import numpy as np
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression # Example model
from sklearn.metrics import accuracy_score # Example metric
from sklearn.datasets import load_iris # Example dataset

In [310]:
# Load sample data
iris = load_iris()
X, y = iris.data, iris.target

In [311]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [312]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [313]:
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['species'] = iris.target
df

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),species
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0
...,...,...,...,...,...
145,6.7,3.0,5.2,2.3,2
146,6.3,2.5,5.0,1.9,2
147,6.5,3.0,5.2,2.0,2
148,6.2,3.4,5.4,2.3,2


In [314]:
df.shape

(150, 5)

In [315]:
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [316]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [268]:
# 1. Initialize KFold
# We choose K=5 folds.
# shuffle=True is recommended to randomize the data order before splitting.
# random_state ensures reproducibility of the shuffle.
k = 5
kf = KFold(n_splits=k, shuffle=True, random_state=42)

fold_accuracies = []
model = LogisticRegression(max_iter=1000) # Instantiate the model once

# 2. Loop through the splits generated by KFold
print(f"Running {k}-Fold Cross-Validation...")
for fold, (train_index, val_index) in enumerate(kf.split(X)):
    # 3. Get training and validation sets for this fold
    X_train, X_val = X[train_index], X[val_index]
    y_train, y_val = y[train_index], y[val_index]

    # 4. Train the model on the training data for this fold
    model.fit(X_train, y_train)

    # 5. Evaluate on the validation data for this fold
    y_pred = model.predict(X_val)
    accuracy = accuracy_score(y_val, y_pred)
    fold_accuracies.append(accuracy)
    print(f"  Fold {fold+1}: Validation Accuracy = {accuracy:.4f}")


Running 5-Fold Cross-Validation...
  Fold 1: Validation Accuracy = 1.0000
  Fold 2: Validation Accuracy = 1.0000
  Fold 3: Validation Accuracy = 0.9333
  Fold 4: Validation Accuracy = 0.9667
  Fold 5: Validation Accuracy = 0.9667


Hocam bu kod bloğunu satır satır açıklayayım.

## `enumerate(kf.split(X))` — İki Kavramı Birleştiren Yapı

Bu ifade, aslında **iki farklı Python/sklearn özelliğinin** iç içe kullanılmış hali. Önce ayrı ayrı, sonra birlikte anlayalım.

### `kf.split(X)` Kısmı (Zaten Bildiğiniz)

Hatırlarsanız bu, her tur için `(train_index, val_index)` çiftini döndüren bir **generator** (üretici) — yani `kf.split(X)`'i bir döngüde gezdirdiğinizde, her adımda size **train ve validation index'lerinden oluşan bir tuple** verir:
```
1. tur: (train_index_1, val_index_1)
2. tur: (train_index_2, val_index_2)
3. tur: (train_index_3, val_index_3)
...
```

### `enumerate()` Kısmı (Yeni Kavram)

`enumerate()`, Python'da **herhangi bir döngüye, otomatik olarak bir "sayaç" (index numarası) ekleyen** bir fonksiyondur. Normalde bir listeyi döngüyle gezerken, kaçıncı elemanda olduğunuzu **elle takip etmeniz** gerekir, ama `enumerate()` bunu **otomatik olarak** size verir:

```python
liste = ['a', 'b', 'c']
for i, harf in enumerate(liste):
    print(i, harf)

# Çıktı:
# 0 a
# 1 b
# 2 c
```

Görüldüğü gibi, `enumerate()` her elemanın yanına, **kaçıncı sırada olduğunu gösteren bir sayı (0'dan başlayarak)** ekliyor.

## İkisini Birleştirdiğimizde — `enumerate(kf.split(X))`

```python
for fold, (train_index, val_index) in enumerate(kf.split(X)):
```

Burada olan şey şu: `kf.split(X)`, her turda `(train_index, val_index)` tuple'ını üretiyor, `enumerate()` ise bu tuple'ların **her birinin önüne**, kaçıncı fold olduğunu gösteren bir **sayaç (0, 1, 2, 3, 4...)** ekliyor. Sonuç olarak her döngü adımında **üç şey** birden elde ediyorsunuz:

```
fold=0: (train_index, val_index)  → 1. fold
fold=1: (train_index, val_index)  → 2. fold
fold=2: (train_index, val_index)  → 3. fold
...
```

`for fold, (train_index, val_index) in ...` yazımındaki **parantez içindeki `(train_index, val_index)`** kısmı, Python'ın **"tuple unpacking" (tuple açma)** özelliği — yani `enumerate` size `(0, (train_idx, val_idx))` gibi **iç içe bir yapı** verirken, siz bunu tek satırda **üç ayrı değişkene** (`fold`, `train_index`, `val_index`) otomatik olarak ayırıyorsunuz.

## Neden `fold` Değişkenine İhtiyaç Var?

Pratik bir sebep var: eğer siz her fold için **ayrı ayrı sonuçlar yazdırmak/kaydetmek** isterseniz (örneğin "1. fold'un accuracy'si şu, 2. fold'unki şu"), `fold` değişkeni size **hangi turda olduğunuzu** söyler:
```python
print(f"Fold {fold+1} - Train boyutu: {len(train_index)}")
```
Bu, sadece **takip/loglama** amaçlı bir kolaylık, K-Fold'un **matematiksel mantığına** hiçbir etkisi yok — sadece kodun **okunabilirliğini ve izlenebilirliğini** artırıyor.

## Şimdi Alt Satırlara Geçelim

```python
X_train, X_val = X[train_index], X[val_index]
y_train, y_val = y[train_index], y[val_index]
```

Bu iki satır, **fancy indexing** dediğimiz bir NumPy/pandas özelliğini kullanıyor — `train_index` ve `val_index`, aslında **bir liste/array of index numaraları** (örneğin `[0, 2, 5, 7, ...]` gibi). `X[train_index]` yazdığınızda, NumPy/pandas bu listeyi alıp, **sadece bu index'lere karşılık gelen satırları** seçip yeni bir array/DataFrame oluşturuyor.

**Somut örnek:**
```python
train_index = [0, 2, 4]
X_train = X[train_index]  # X'in 0., 2., 4. satırlarını seçer
```

Burada dikkat edilmesi gereken nokta: **önceki örneklerde** (`df.iloc[train_index]`) pandas DataFrame kullanıyorduk, `.iloc[]` gerekliydi. Bu kodda ise `X[train_index]` direkt köşeli parantez kullanılmış — bu, `X`'in muhtemelen bir **NumPy array** olduğunu gösteriyor (Iris veri setini `load_iris()` ile çektiğinizde, `iris.data` zaten NumPy array formatında gelir, pandas DataFrame değil). NumPy array'lerde direkt `X[liste]` kullanmak geçerlidir, pandas DataFrame'lerde ise `.iloc[]` kullanmanız gerekirdi — bu, veri tipine göre değişen küçük ama önemli bir syntax farkı.

## Neden Hem `X` Hem `y` İçin Aynı İşlemi Yapıyoruz?

Çünkü bir modeli eğitmek için hem **özellikler (X)** hem **hedef değişken (y)** gerekiyor, ve bunların **birbirleriyle satır satır eşleşmesi** kritik — yani `X_train`'in 5. satırı ile `y_train`'in 5. satırı, **aynı çiçeğe/gözleme ait** olmalı. `train_index` ve `val_index` **aynı index listeleri** olduğu için, hem `X`'i hem `y`'yi **bu aynı index'lerle** bölerek, aralarındaki **eşleşmeyi bozmadan** train/validation ayrımını yapmış oluyorsunuz.

---

## Tüm Kodu Bir Cümlede Özetleyelim

Bu kod, K-Fold'un ürettiği her `(train_index, val_index)` çiftini, **hangi fold olduğunu gösteren bir sayaçla (`fold`) birlikte** döngüye alıyor, sonra bu index listelerini kullanarak hem özellik matrisini (`X`) hem hedef değişkeni (`y`), **birbirleriyle tutarlı bir şekilde**, train ve validation olmak üzere ikiye ayırıyor — bu, az önce anlattığımız K-Fold'un **genel prensibinin**, gerçek bir model eğitimi döngüsünde **nasıl koda döküldüğünün** somut bir örneği.

Hocam sorun değil, bunu somut bir örnekle, sıfırdan açayım — muhtemelen kafanızı karıştıran şey, **"X zaten bölündü, neden y'yi de ayrıca bölüyoruz"** sorusu olabilir.

## Önce X ve y'nin Ne Olduğunu Hatırlayalım

ML'de veri setiniz genelde **ikiye ayrılır**:
- **X (features/özellikler)** → modelin **tahmin yapmak için kullanacağı** girdiler (örneğin Iris'te: yaprak uzunluğu, genişliği gibi 4 sütun)
- **y (target/hedef)** → modelin **tahmin etmeye çalıştığı** çıktı (örneğin Iris'te: çiçeğin türü — Setosa/Versicolor/Virginica)

Bunlar, veri setinde **iki ayrı, bağımsız değişken/array** olarak tutulur:
```python
X = iris.data      # şekli: (150, 4) → 150 satır, 4 sütun (özellikler)
y = iris.target    # şekli: (150,)   → 150 satır, tek sütun (hedef)
```

**Kritik nokta:** `X` ve `y`, **fiziksel olarak iki ayrı Python nesnesidir** — aynı DataFrame/array içinde birleşik değiller, iki **ayrı** değişken. Bu yüzden birini bölmek, **diğerini otomatik olarak bölmez** — ikisini de **elle, ayrı ayrı** bölmeniz gerekir.

## Somut Bir Örnekle Gösterelim

Diyelim ki elimizde sadece 5 çiçek var (basitleştirelim):

```python
X = [[5.1, 3.5], [4.9, 3.0], [6.2, 2.8], [5.5, 2.3], [7.0, 3.2]]  # 5 satır, 2 özellik
y = [0, 0, 1, 1, 2]  # her satırın türü (0=Setosa, 1=Versicolor, 2=Virginica)
```

Burada **X'in 3. satırı** (`[6.2, 2.8]`) ile **y'nin 3. elemanı** (`1`) **birbirine karşılık geliyor** — yani bu, **"6.2, 2.8 ölçümlerine sahip çiçek, Versicolor türüdür"** demek. Bu **eşleşme**, index numaraları üzerinden kuruluyor.

## Şimdi `train_index = [0, 2, 4]` Olduğunu Varsayalım

```python
X_train = X[train_index]  # X'in 0., 2., 4. satırlarını al
# X_train = [[5.1, 3.5], [6.2, 2.8], [7.0, 3.2]]

y_train = y[train_index]  # y'nin 0., 2., 4. elemanlarını al
# y_train = [0, 1, 2]
```

**İşte kritik nokta burada:** `X_train`'in **1. satırı** (`[5.1, 3.5]`) ile `y_train`'in **1. elemanı** (`0`) **hâlâ birbirine karşılık geliyor** — yani "5.1, 3.5 ölçümlü çiçek, Setosa'dır" bilgisi **korunmuş**. Eğer siz **sadece X'i bölüp, y'yi bölmeseydiniz**, `y` hâlâ **eski, 5 elemanlı haliyle** kalırdı, ama `X_train`'iniz artık **sadece 3 satırlı**, bu ikisi **artık uyumsuz** olurdu — model, `X_train`'in 1. satırını `y`'nin **hangi** elemanıyla eşleştireceğini bilemezdi, çünkü **boyutlar tutmuyor** olurdu (X_train 3 satır, y hâlâ 5 eleman).

## Neden Bu Kod Satırının Ayrı Olması Gerekiyor?

```python
X_train, X_val = X[train_index], X[val_index]
y_train, y_val = y[train_index], y[val_index]
```

Bu **iki ayrı satır**, çünkü Python'da `X` ve `y` **birbirinden habersiz, bağımsız iki nesne** — birini index'lemek, diğerini **otomatik olarak** index'lemez. Siz **aynı `train_index` ve `val_index` listelerini**, hem `X`'e hem `y`'ye **ayrı ayrı ama tutarlı bir şekilde** uyguluyorsunuz, böylece:

- `X_train` ve `y_train` **birbirleriyle eşleşmiş** kalıyor (model eğitiminde kullanılacak)
- `X_val` ve `y_val` **birbirleriyle eşleşmiş** kalıyor (model test edilirken kullanılacak)

## Bir Benzetmeyle Pekiştirelim

Bunu bir **sınıf listesi ve not listesi** gibi düşünün — elinizde iki ayrı liste var:
```
Öğrenci_isimleri = ["Ali", "Veli", "Ayşe", "Fatma", "Mehmet"]
Öğrenci_notları  = [85,     92,     78,     90,      65]
```
Eğer siz "ilk 3 öğrenciyi seç" derseniz, **hem isim listesinden hem not listesinden**, **aynı 3 index'i (0, 1, 2)** almanız gerekir — sadece isimleri seçip notları unutursanız, hangi notun hangi öğrenciye ait olduğunu **kaybedersiniz**. `X` ve `y`'yi ayrı ayrı bölmek, tam olarak bu **"isim ve not listesini birlikte, tutarlı şekilde seçme"** işlemine karşılık geliyor.

---

**Özetle:** `X` ve `y`, Python'da **iki ayrı, bağımsız nesne** olduğu için, birini bölmek **diğerini otomatik olarak bölmez** — bu yüzden **aynı `train_index`/`val_index` listelerini**, hem `X`'e hem `y`'ye **ayrı ayrı** uygulamanız gerekiyor. Bu, ikisinin **satır satır eşleşmesini** (yani "bu özellikler, bu hedefe ait" bilgisini) **bozmadan** korumanın tek yolu — aksi halde, model hangi girdi ile hangi çıktının eşleştiğini bilemez hale gelirdi.

In [269]:
# 6. Calculate average performance and standard deviation
mean_accuracy = np.mean(fold_accuracies)
std_accuracy = np.std(fold_accuracies)

print(f"\nCross-Validation Results ({k}-Fold):")
print(f"  Individual Fold Accuracies: {[f'{acc:.4f}' for acc in fold_accuracies]}")
print(f"  Average Validation Accuracy: {mean_accuracy:.4f}")
print(f"  Standard Deviation of Accuracy: {std_accuracy:.4f}")

# Expected Output:
# Running 5-Fold Cross-Validation...
#   Fold 1: Validation Accuracy = 1.0000
#   Fold 2: Validation Accuracy = 0.9667
#   Fold 3: Validation Accuracy = 0.9333
#   Fold 4: Validation Accuracy = 0.9333
#   Fold 5: Validation Accuracy = 0.9667
#
# Cross-Validation Results (5-Fold):
#   Individual Fold Accuracies: ['1.0000', '0.9667', '0.9333', '0.9333', '0.9667']
#   Average Validation Accuracy: 0.9600
#   Standard Deviation of Accuracy: 0.0249



Cross-Validation Results (5-Fold):
  Individual Fold Accuracies: ['1.0000', '1.0000', '0.9333', '0.9667', '0.9667']
  Average Validation Accuracy: 0.9733
  Standard Deviation of Accuracy: 0.0249


# In this code:

We initialize KFold with n_splits=5, shuffle=True, and random_state=42. Shuffling is generally a good idea unless your data has inherent sequential order that you need to preserve. random_state makes the shuffle predictable, which is important for getting reproducible results.

The kf.split(X) method generates pairs of index arrays (train_index, val_index) for each fold.

Inside the loop, we use these indices to slice X and y into the appropriate training and validation sets for the current fold.

We train a LogisticRegression model and evaluate its accuracy_score on the validation set.

After the loop, we compute the mean and standard deviation of the accuracies obtained across the 5 folds. The mean accuracy (0.9600 here) gives us our primary estimate of the model's performance. The standard deviation (0.0249) tells us about the variability of the performance across different folds; a lower standard deviation suggests more consistent performance.

# Using cross_val_score for Convenience

Manually looping through folds is instructive, but Scikit-learn provides the cross_val_score function, which performs the entire K-Fold cross-validation process (splitting, training, evaluating) much more concisely.

(Note: If cv is an integer and the estimator is a classifier, cross_val_score uses StratifiedKFold by default, which we'll discuss next. For regression or if shuffle=False, it uses KFold. The slight difference in average accuracy compared to the manual example might be due to this or minor implementation details. To exactly replicate the manual KFold behavior with shuffle, you can pass the kf object directly: scores = cross_val_score(model, X, y, cv=kf, scoring='accuracy'))

The cross_val_score function takes the estimator (your model instance), features X, target y, the cross-validation strategy cv (here, just the integer 5, meaning use 5 folds), and a scoring metric string. It returns an array containing the score for each fold. This is significantly more compact than the manual loop.

Important Note on Scoring: Scikit-learn's scoring functions follow a convention where higher values are better. For metrics like Mean Squared Error (MSE) or Mean Absolute Error (MAE), where lower values are better, Scikit-learn provides negated versions like 'neg_mean_squared_error' or 'neg_mean_absolute_error'. When using cross_val_score with these, you'll get negative scores, and a "better" result will be closer to zero (i.e., a smaller negative number).

Scikit-learn'ün skorlama fonksiyonları, "yüksek değer her zaman daha iyidir" kuralına göre tasarlanmıştır; yani `cross_val_score` gibi fonksiyonlar, hangi metriği kullanırsanız kullanın, **her zaman büyük sayının kazandığı** bir mantıkla çalışır — accuracy için bu doğaldır, çünkü accuracy zaten "%90, %80'den iyidir" şeklinde yüksek-iyi mantığına uyar. Ama Mean Squared Error (MSE) ya da Mean Absolute Error (MAE) gibi hata metriklerinde bu mantık **tam tersine döner** — bu metriklerde **düşük değer** daha iyidir, çünkü MSE/MAE, modelin tahminlerinin gerçek değerlerden **ne kadar saptığını** ölçer, ve elbette az sapma (küçük sayı) iyi, çok sapma (büyük sayı) kötüdür. İşte tam burada bir çelişki ortaya çıkıyor: scikit-learn'ün "büyük sayı iyi" kuralı ile MSE/MAE'nin "küçük sayı iyi" doğası birbiriyle **uyuşmuyor**. Scikit-learn bu çelişkiyi çözmek için basit ama zekice bir hile kullanıyor: MSE/MAE değerlerinin **negatifini** alarak sunuyor, yani `'neg_mean_squared_error'` gibi bir isimle, hesaplanan MSE değerini **eksi işaretiyle** çarpıp döndürüyor. Bunun mantığı şu: eğer gerçek MSE değerleri `[100, 50, 10]` gibi çıkıyorsa (10 en iyi, 100 en kötü), bunların negatifini aldığınızda `[-100, -50, -10]` elde edersiniz — ve şimdi **-10, -100'den büyüktür**, yani "büyük sayı iyi" kuralı **yeniden geçerli hale gelir**, çünkü -10 (orijinalde en düşük/en iyi MSE olan 10'un negatifi) artık listedeki **en büyük** sayıdır. Bu sayede scikit-learn, MSE/MAE gibi "az olması iyi" metrikleri de, kendi tutarlı "çok olması iyi" sistemine **zorlamadan, matematiksel bir işaret değişikliğiyle** uydurmuş oluyor. Pratikte bu, `cross_val_score(model, X, y, scoring='neg_mean_squared_error')` kullandığınızda, çıkan sonuçların **hep negatif** göründüğü, ve "en iyi" sonucun **sıfıra en yakın** (yani en küçük mutlak değere sahip) negatif sayı olduğu anlamına gelir — örneğin `-5`, `-50`'den **daha iyi bir sonuçtur**, çünkü mutlak değeri daha küçüktür, yani gerçek hata daha azdır; bu yüzden bu skorları yorumlarken kafanızda otomatik olarak **işareti geri çevirip** ("bu aslında pozitif 5 MSE demek") düşünmeniz gerekir.

By using K-Fold cross-validation, either manually or with cross_val_score, you obtain a more trustworthy estimate of how your model is likely to perform on new, unseen data compared to a single train-test split. This is a fundamental technique for reliable model evaluation and comparison.

In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.datasets import load_iris

# Load sample data
iris = load_iris()
X, y = iris.data, iris.target

# Instantiate the model
model = LogisticRegression(max_iter=1000)

# Define the number of folds (K)
k = 5

# Use cross_val_score
# cv can be an integer (for KFold or StratifiedKFold) or a CV splitter object
# scoring specifies the metric to use (e.g., 'accuracy', 'neg_mean_squared_error')
scores = cross_val_score(model, X, y, cv=k, scoring='accuracy')

print(f"Running {k}-Fold Cross-Validation using cross_val_score...")
print(f"  Scores for each fold: {scores}")
print(f"  Average Accuracy: {scores.mean():.4f}")
print(f"  Standard Deviation: {scores.std():.4f}")

# Expected Output (may differ slightly from manual loop if shuffle isn't explicitly matched):
# Running 5-Fold Cross-Validation using cross_val_score...
#   Scores for each fold: [0.96666667 1.         0.93333333 0.96666667 1.        ]
#   Average Accuracy: 0.9733
#   Standard Deviation: 0.0249


In [270]:
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import numpy as np

kf = KFold(n_splits=5, shuffle=True, random_state=42)
fold_accuracies = []  # her fold'un skorunu biriktireceğimiz liste

for fold, (train_index, val_index) in enumerate(kf.split(X)):
    X_train, X_val = X[train_index], X[val_index]
    y_train, y_val = y[train_index], y[val_index]
    
    # 1. Modeli oluştur
    model = LogisticRegression(max_iter=1000)
    
    # 2. SADECE bu fold'un train verisiyle eğit
    model.fit(X_train, y_train)
    
    # 3. Bu fold'un validation verisinde tahmin yap
    y_pred = model.predict(X_val)
    
    # 4. Bu fold'un doğruluğunu hesapla
    acc = accuracy_score(y_val, y_pred)
    fold_accuracies.append(acc)
    
    print(f"Fold {fold+1} - Accuracy: {acc:.4f}")

print("\nOrtalama Accuracy:", np.mean(fold_accuracies))
print("Standart Sapma:", np.std(fold_accuracies))

Fold 1 - Accuracy: 1.0000
Fold 2 - Accuracy: 1.0000
Fold 3 - Accuracy: 0.9333
Fold 4 - Accuracy: 0.9667
Fold 5 - Accuracy: 0.9667

Ortalama Accuracy: 0.9733333333333334
Standart Sapma: 0.024944382578492935


In [317]:
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import numpy as np

kf = KFold(n_splits=5, shuffle=True, random_state=42)
fold_accuracies = []  # her fold'un skorunu biriktireceğimiz liste

for fold, (train_index, val_index) in enumerate(kf.split(X)):
    X_train, X_val = X[train_index], X[val_index]
    y_train, y_val = y[train_index], y[val_index]
    
    # 1. Modeli oluştur
    model = LogisticRegression(max_iter=1000)
    
    # 2. SADECE bu fold'un train verisiyle eğit
    model.fit(X_train, y_train)
    
    # 3. Bu fold'un validation verisinde tahmin yap
    y_pred = model.predict(X_val)
    
    # 4. Bu fold'un doğruluğunu hesapla
    acc = accuracy_score(y_val, y_pred)
    fold_accuracies.append(acc)
    
    print(f"Fold {fold+1} - Accuracy: {acc:.4f}")

print("\nOrtalama Accuracy:", np.mean(fold_accuracies))
print("Standart Sapma:", np.std(fold_accuracies))

Fold 1 - Accuracy: 1.0000
Fold 2 - Accuracy: 1.0000
Fold 3 - Accuracy: 0.9333
Fold 4 - Accuracy: 0.9667
Fold 5 - Accuracy: 0.9667

Ortalama Accuracy: 0.9733333333333334
Standart Sapma: 0.024944382578492935


## Yeni Eklenen Kısımları Açıklayalım

**1. Neden Modeli Döngünün İÇİNDE Oluşturuyoruz?**
```python
model = LogisticRegression(max_iter=1000)
```
Bu satırın **her fold'da tekrar** çalışması **kritik** — çünkü her fold'da modeli **sıfırdan, temiz bir şekilde** eğitmek istiyoruz. Eğer modeli döngünün **dışında bir kere** oluşturup, her fold'da **aynı modeli** tekrar tekrar `fit()` etseydiniz, model **önceki fold'ların bilgisini de taşımaya devam ederdi** — bu da fold'ların **birbirinden bağımsız** olması gereken temel prensibi bozardı.

**2. `fold_accuracies.append(acc)` — Sonuçları Biriktirme**
Her fold'un skorunu bir **listeye** ekliyoruz, böylece döngü bittiğinde elimizde **5 farklı accuracy değeri** oluyor — tek bir sayı değil, **5 sayıdan oluşan bir dağılım**.

**3. Neden Sadece Ortalama Değil, Standart Sapma da Hesaplıyoruz?**
```python
np.std(fold_accuracies)
```
Bu, size modelin **ne kadar tutarlı** olduğunu gösterir. Eğer 5 fold'un accuracy'leri birbirine çok yakınsa (örneğin hepsi 0.78-0.80 arası), **düşük standart sapma** çıkar, bu da modelin **kararlı, güvenilir** olduğunu gösterir. Eğer fold'lar arasında büyük farklar varsa (örneğin biri 0.60, biri 0.90), **yüksek standart sapma** çıkar — bu da modelin **veri setinin hangi kısmına denk geldiğine göre** çok değişken performans gösterdiğini, yani **güvenilmez** olduğunu işaret eder.

## Bunu Otomatikleştiren Hazır Fonksiyon — `cross_val_score`

Sklearn, bu **tüm döngüyü** sizin yerinize yapan hazır bir fonksiyon sunuyor:


**Bu fonksiyon, arka planda tam olarak** sizin manuel yazdığınız döngüyü **otomatik olarak** çalıştırıyor — 5 kez model eğitip, 5 kez test edip, sonuçları bir array olarak döndürüyor.

## Peki Neden Önce Manuel Yazmayı Öğrendik, Hazır Fonksiyon Varken?

Bu, gerçekten önemli bir soru — cevap şu: **`cross_val_score`, sadece "model performansı ölçmek" için kullanışlı**, ama **Target Encoding'de yaptığımız gibi özel bir işlem** (her fold'da farklı bir hesaplama yapıp, sonucu ana veri setine geri yazmak) gerektiğinde, **hazır fonksiyon yetmez**, **manuel döngü yazmanız gerekir** — tam da bizim Target Encoding kodunda yaptığımız gibi. Yani:

- **Sadece model performansı ölçmek istiyorsanız** → `cross_val_score` yeterli ve pratik
- **Fold'ların içinde özel bir işlem (encoding, custom preprocessing) yapmanız gerekiyorsa** → manuel döngü şart

## Pratik Bir Kıyaslama

| | Manuel Döngü | `cross_val_score` |
|---|---|---|
| Kod uzunluğu | Uzun | Kısa (2-3 satır) |
| Esneklik | Tam kontrol, her adımı özelleştirebilirsiniz | Sınırlı, sadece standart model değerlendirmesi |
| Ne zaman kullanılır | Target Encoding gibi özel işlemler | Basit model performans karşılaştırması |

---

**Özetle:** K-Fold döngüsünün geri kalanı, her fold'da **modeli sıfırdan eğitip**, **validation setinde test edip**, sonuçları **biriktirip ortalamasını almaktan** ibaret — bu, size modelin **tek bir bölünmeye bağlı olmayan, daha güvenilir bir performans tahmini** verir. Sklearn'ün `cross_val_score` fonksiyonu bu süreci **otomatikleştirir**, ama Target Encoding gibi **fold-içi özel işlemler** gerektiğinde, sizin öğrendiğiniz **manuel döngü yaklaşımı** vazgeçilmez oluyor.

In [271]:
from sklearn.model_selection import cross_val_score

model = LogisticRegression(max_iter=1000)
scores = cross_val_score(model, X, y, cv=kf)

print("Her fold'un accuracy'si:", scores)
print("Ortalama:", scores.mean())
print("Standart Sapma:", scores.std())

Her fold'un accuracy'si: [1.         1.         0.93333333 0.96666667 0.96666667]
Ortalama: 0.9733333333333334
Standart Sapma: 0.024944382578492935


Soru: Hocam bu modeli biz neden dışarıda yazdık. Bunu daha da detaylandırır  mısın ?

Cevap:

Hocam çok iyi bir gözlem, gerçekten kafa karıştırıcı bir tutarsızlık gibi görünüyor — çünkü ben az önce **"model her fold'da içeride, sıfırdan oluşturulmalı"** dedim, ama `cross_val_score` örneğinde modeli **döngünün dışında, bir kere** yazdık. Bu çelişkiyi açıklığa kavuşturayım.

## Görünürdeki Çelişki

```python
# Manuel döngüde - modeli İÇERİDE, her fold'da yeniden oluşturduk
for fold, (train_index, val_index) in enumerate(kf.split(X)):
    model = LogisticRegression(max_iter=1000)  # HER FOLD'DA yeni model
    model.fit(X_train, y_train)

# cross_val_score'da - modeli DIŞARIDA, bir kere oluşturduk
model = LogisticRegression(max_iter=1000)  # SADECE BİR KERE
scores = cross_val_score(model, X, y, cv=kf)
```

**Soru şu: `cross_val_score`'a verdiğimiz bu "dışarıdaki" model, 5 fold boyunca aynı model olarak mı kalıyor, yoksa her fold'da temiz bir başlangıç mı yapıyor?**

## Cevap — `cross_val_score` Arka Planda "Klonlama" Yapıyor

İşte kritik nokta burada: siz `cross_val_score`'a bir model **nesnesi** veriyorsunuz, ama `cross_val_score` bu nesneyi **doğrudan kullanıp tekrar tekrar `fit()` etmiyor**. Bunun yerine, sklearn'ün içinde **`clone()`** dediğimiz bir mekanizma var — her fold başladığında, `cross_val_score` sizin verdiğiniz modelin **parametrelerini kopyalayarak** (yani `max_iter=1000` gibi ayarları alarak), **tamamen yeni, hiç eğitilmemiş bir model kopyası** oluşturuyor.

```python
# cross_val_score'un ARKA PLANDA yaptığı şey, kabaca şuna benzer:
from sklearn.base import clone

for fold, (train_index, val_index) in enumerate(kf.split(X)):
    model_kopyasi = clone(model)  # sizin verdiğiniz modelin "temiz" bir kopyası
    model_kopyasi.fit(X_train, y_train)
    # ...
```

Yani siz `model = LogisticRegression(max_iter=1000)` yazdığınızda, aslında **eğitilmiş bir model değil**, sadece **"şu ayarlarla bir model kur" talimatını taşıyan bir şablon (template/blueprint)** oluşturuyorsunuz. `cross_val_score`, bu şablonu **her fold için ayrı ayrı, sıfırdan** kullanıyor — modelin **kendisini** değil, **tarifini/ayarlarını** paylaşıyorsunuz.

## Neden Bu Tasarım Tercih Edilmiş?

Bu, **pratiklik ve okunabilirlik** için yapılan bilinçli bir tasarım kararı — sklearn, kullanıcıların **her seferinde "her fold'da yeni model oluştur" diye elle yazmak zorunda kalmasını istemiyor**, bu yüzden `cross_val_score` bu **klonlama işini sizin yerinize, görünmez bir şekilde** hallediyor. Siz sadece **"bu tür bir model, bu ayarlarla"** diyorsunuz, `cross_val_score` da **"tamam, bu tarife göre 5 farklı, birbirinden bağımsız model kuracağım"** diyor.

## Somut Bir Kanıtla Gösterelim

```python
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

model = LogisticRegression(max_iter=1000)
print("Eğitim öncesi model eğitilmiş mi?:", hasattr(model, 'coef_'))  # False - henüz eğitilmemiş

scores = cross_val_score(model, X, y, cv=kf)

print("cross_val_score sonrası model eğitilmiş mi?:", hasattr(model, 'coef_'))  # HALA False!
```

**İşte kanıt:** `cross_val_score` çalıştıktan **sonra bile**, sizin elinizdeki `model` değişkeni **hâlâ eğitilmemiş** durumda (`coef_` attribute'u yok) — çünkü `cross_val_score`, sizin `model` nesnenizi **hiç `fit()` etmedi**, sadece onun **klonlarını** kullandı ve o klonları eğitti, sonra attı. Sizin orijinal `model` nesneniz **hep "bakir" (el değmemiş)** kaldı.

## Manuel Döngüyle Karşılaştıralım — Neden Orada Farklı Davrandık?

Manuel döngüde, biz **kendi elimizle** `model.fit()` çağırdığımız için, eğer modeli **döngünün dışında bir kere** yazsaydık:

```python
model = LogisticRegression(max_iter=1000)  # DIŞARIDA

for fold, ... in enumerate(kf.split(X)):
    model.fit(X_train, y_train)  # AYNI model nesnesini tekrar tekrar eğitiyoruz!
```

Bu durumda **gerçek bir sorun olurdu** — çünkü `model.fit()` her çağrıldığında, **var olan modelin üzerine yazar** (sklearn modelleri, `fit()` her çağrıldığında **önceki eğitimi sıfırlar ve yeniden başlar**, ama yine de bu **kötü bir pratik** olurdu, çünkü kodun okunurluğu ve niyeti bulanıklaşır, ayrıca bazı model türlerinde/durumlarda beklenmedik yan etkiler olabilir). Bu yüzden **manuel döngüde, açıkça ve güvenli olması için modeli içeride yazmak** doğru pratiktir — sklearn'ün otomatik klonlama gibi bir "sigortası" olmadığı için, siz **kendi sigortanızı kendiniz** kurmalısınız.

## Özetle İki Yaklaşımın Farkı

| | Manuel Döngü | `cross_val_score` |
|---|---|---|
| Model nereye yazılır | Döngünün **içine** | Döngünün **dışına** (çünkü döngü zaten görünmez) |
| Neden | Sizin `fit()` çağrınız, **gerçek** modeli eğitiyor, temiz başlangıç için içeride olmalı | `cross_val_score` **otomatik klonlama** yaptığı için, dışarıdaki model sadece bir "şablon" |
| Risk | İçeride yazmazsanız, aynı model tekrar tekrar eğitilir (mantık hatası) | Dışarıda yazmak sorun değil, çünkü zaten hiç doğrudan kullanılmıyor |

---

**Özetle:** `cross_val_score`'da modeli dışarıda yazmamızın sebebi, bu fonksiyonun **arka planda otomatik olarak modeli klonlayıp**, her fold için **temiz, bağımsız bir kopya** kullanmasıdır — siz sadece **"nasıl bir model istiyorum"** tarifini veriyorsunuz, gerçek eğitim/klonlama işini `cross_val_score` sizin yerinize yapıyor. Manuel döngüde ise böyle bir otomatik klonlama mekanizması olmadığı için, **siz** modeli her fold'da **elle, içeride** yeniden oluşturmak zorundasınız — aksi halde aynı model nesnesi tekrar tekrar `fit()` edilir ki bu, mantıksal olarak istediğimiz **"her fold bağımsız olsun"** prensibini bulanıklaştırır.

# Stratified K-Fold Cross Validation

Scikit-learn makes using Stratified K-Fold straightforward. You can use the StratifiedKFold cross-validator explicitly or rely on functions like cross_val_score which often use stratification automatically for classification tasks.

Let's see how to use StratifiedKFold explicitly and then integrate it with cross_val_score.

In [320]:
import numpy as np
from sklearn.datasets import make_classification
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.linear_model import LogisticRegression

# Create a synthetic imbalanced dataset
X, y = make_classification(n_samples=100, n_features=20, n_informative=2,
                           n_redundant=10, n_clusters_per_class=1,
                           weights=[0.9, 0.1], flip_y=0, random_state=42)

print(f"Dataset shape: {X.shape}")
print(f"Class distribution: {np.bincount(y)}")

# Define the classifier
classifier = LogisticRegression(solver='liblinear', random_state=42)


Dataset shape: (100, 20)
Class distribution: [90 10]


Now, let's explicitly create a StratifiedKFold object and see how it generates indices:

In [274]:
# Initialize Stratified K-Fold
n_splits = 5
skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=42)

# Iterate through the splits and show indices (optional: just for illustration)
print(f"\nGenerating {n_splits} stratified folds:")
fold_num = 1
for train_index, test_index in skf.split(X, y):
    print(f"Fold {fold_num}:")
    # print(f"  Train indices: {train_index[:10]}...") # Optional: view indices
    # print(f"  Test indices:  {test_index}")       # Optional: view indices
    print(f"  Train size: {len(train_index)}, Test size: {len(test_index)}")
    # Verify class distribution in the test set for this fold
    y_test_fold = y[test_index]
    print(f"  Test fold class distribution: {np.bincount(y_test_fold)}")
    fold_num += 1



Generating 5 stratified folds:
Fold 1:
  Train size: 80, Test size: 20
  Test fold class distribution: [18  2]
Fold 2:
  Train size: 80, Test size: 20
  Test fold class distribution: [18  2]
Fold 3:
  Train size: 80, Test size: 20
  Test fold class distribution: [18  2]
Fold 4:
  Train size: 80, Test size: 20
  Test fold class distribution: [18  2]
Fold 5:
  Train size: 80, Test size: 20
  Test fold class distribution: [18  2]


You'll notice that each test fold maintains approximately the same ratio of classes as the original dataset (around 90% and 10%).

The most common way to use Stratified K-Fold is by passing it (or simply the number of splits, letting Scikit-learn choose stratification) to cross_val_score:

In [ ]:
# Method 1: Pass the StratifiedKFold object directly
scores_explicit_skf = cross_val_score(classifier, X, y, cv=skf, scoring='accuracy')
print(f"\nAccuracy scores using explicit StratifiedKFold ({n_splits}-fold): {scores_explicit_skf}")
print(f"Mean Accuracy: {scores_explicit_skf.mean():.4f} (+/- {scores_explicit_skf.std() * 2:.4f})")

# Method 2: Pass an integer - cross_val_score often defaults to StratifiedKFold for classifiers
# Note: For reproducibility when relying on the default, ensure the estimator has random_state set if applicable,
# and consider that the internal shuffling might differ if not explicitly controlled via a cv object.
scores_implicit_skf = cross_val_score(classifier, X, y, cv=n_splits, scoring='accuracy')
print(f"\nAccuracy scores using implicit stratification ({n_splits}-fold): {scores_implicit_skf}")
print(f"Mean Accuracy: {scores_implicit_skf.mean():.4f} (+/- {scores_implicit_skf.std() * 2:.4f})")



Accuracy scores using explicit StratifiedKFold (5-fold): [1.   1.   0.95 1.   1.  ]
Mean Accuracy: 0.9900 (+/- 0.0400)

Accuracy scores using implicit stratification (5-fold): [1.   1.   0.95 1.   1.  ]
Mean Accuracy: 0.9900 (+/- 0.0400)


Both methods achieve the same goal: evaluating the classifier using cross-validation where each fold preserves the overall class balance.

When to Use Stratified K-Fold:

Classification Tasks: It's generally the recommended default for classification.

Imbalanced Datasets: It's particularly important when dealing with datasets where one class significantly outnumbers others.

Small Datasets: Stratification helps ensure that rare classes are represented in all folds, which is critical when data is limited.

For regression problems, standard K-Fold is typically sufficient, as the concept of "class balance" doesn't directly apply. However, for classification, using Stratified K-Fold provides a more reliable and stable estimate of your model's performance on unseen data, helping you make better decisions during model selection and hyperparameter tuning.

In [277]:
# Example using scikit-learn for Stratified K-Fold
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.ensemble import RandomForestClassifier
import numpy as np

# Assume X and y are defined, y might be imbalanced
# Example placeholder imbalanced data
X = np.random.rand(100, 10)
y = np.array([0]*80 + [1]*20) # 80% class 0, 20% class 1
np.random.shuffle(y) # Shuffle to mix classes

model = RandomForestClassifier(random_state=42)

# Define the Stratified K-Fold strategy (e.g., 5 folds)
stratified_kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation using Stratified K-Fold
# For classification, cross_val_score often uses StratifiedKFold by default
# if y is passed, but specifying it explicitly is clearer.
scores = cross_val_score(model, X, y, cv=stratified_kfold, scoring='roc_auc') # Using ROC AUC

print(f"Stratified Scores for each fold: {scores}")
print(f"Average CV ROC AUC: {np.mean(scores):.4f}")
print(f"Standard Deviation of CV ROC AUC: {np.std(scores):.4f}")

Stratified Scores for each fold: [0.3125    0.3828125 0.390625  0.65625   0.421875 ]
Average CV ROC AUC: 0.4328
Standard Deviation of CV ROC AUC: 0.1173


Tabii hocam, sorayım demeden başlayalım.

## `scoring` parametresi nedir?

`cross_val_score` her fold'da modeli train edip validation set üzerinde bir performans skoru hesaplar. Bu skorun **hangi metrikle** hesaplanacağını `scoring` parametresi belirler.

Varsayılan davranış: `scoring` belirtilmezse, model'in kendi `.score()` metodu kullanılır — classifier için bu genelde **accuracy**'dir. Ama accuracy her zaman doğru metrik değildir (özellikle dengesiz veri setlerinde), o yüzden `scoring='roc_auc'` gibi açıkça belirtmen gerekir.

sklearn'de `scoring` bir string alabilir (`'accuracy'`, `'f1'`, `'roc_auc'`, `'neg_mean_squared_error'` gibi) ya da özel bir scorer fonksiyonu (`make_scorer` ile) alabilir.

## ROC AUC nedir?

Bunu iki parçaya ayıralım: **ROC eğrisi** ve **AUC**.

### ROC (Receiver Operating Characteristic) Eğrisi

Bir binary classifier aslında bir olasılık üretir (örn. "bu hasta %73 ihtimalle pozitif"), sonra bunu bir **threshold** (genelde 0.5) ile 0/1'e çevirirsin. Ama bu threshold'u değiştirirsen model'in davranışı da değişir:

- Threshold düşürürsen → daha çok şeyi "pozitif" dersin → **True Positive Rate (Recall)** artar ama **False Positive Rate** de artar
- Threshold yükseltirsen → tam tersi

ROC eğrisi, **her mümkün threshold** için bu iki değeri (TPR vs FPR) çizer:

- **X ekseni:** False Positive Rate = `FP / (FP + TN)` — yanlışlıkla pozitif dediklerinin oranı
- **Y ekseni:** True Positive Rate = `TP / (TP + FN)` — gerçek pozitifleri yakalama oranı (Recall ile aynı şey)

### AUC (Area Under Curve)

ROC eğrisinin altındaki alan. Tek bir sayıya indirger:

- **AUC = 1.0** → mükemmel ayrım (model pozitif ve negatifleri hatasız ayırıyor)
- **AUC = 0.5** → rastgele tahmin (coin flip) ile aynı, model hiçbir şey öğrenmemiş
- **AUC < 0.5** → model'in tahminleri ters çevrilirse daha iyi olur (bir şeyler ciddi ters gitmiş demektir)

**Sezgisel yorum:** AUC = 0.85 demek, rastgele seçilen bir pozitif örneğin, rastgele seçilen bir negatif örnekten daha yüksek skor alma olasılığı %85'tir.

### Neden accuracy yerine ROC AUC?

Bu senin data'nda da geçen bir konu — **fraud detection gibi dengesiz veri setlerinde** kritik:

Diyelim 1000 hastadan 10'u hasta (%1 pozitif oranı). Model her şeye "sağlıklı" dese bile **%99 accuracy** alır — ama hiçbir işe yaramaz. ROC AUC ise:

- **Threshold'dan bağımsızdır** — tüm threshold'ları değerlendirir, tek bir eşiğe bağlı kalmaz
- **Sınıf dengesizliğine accuracy kadar duyarlı değildir** — modelin sıralama (ranking) gücünü ölçer, yani "pozitifleri negatiflerden ne kadar iyi ayırt ediyor" sorusuna cevap verir

### Küçük bir uyarı

ROC AUC da mükemmel değil — **aşırı dengesiz** veri setlerinde (örn. %0.1 pozitif oranı) bazen yanıltıcı olabilir, çünkü FPR hesabında paydadaki TN sayısı çok büyük olduğu için FPR yapay olarak düşük kalır. Böyle aşırı durumlarda **Precision-Recall AUC** (`scoring='average_precision'`) daha bilgilendirici olabilir.

Senin kod örneğindeki kullanım tam yerinde: `StratifiedKFold` ile fold'larda sınıf oranını koruyorsun, `scoring='roc_auc'` ile de accuracy tuzağına düşmeden gerçek ayırt edici gücü ölçüyorsun.

## Şu an gerçekten ihtiyacın olan (ezberle, bitti)

Pratikte %90 ihtimalle şu 4-5 tanesini kullanacaksın:

| Problem tipi | scoring string'i | Ne zaman |
|---|---|---|
| Classification (dengeli) | `'accuracy'` | Sınıflar eşit dağılımlıysa |
| Classification (dengesiz) | `'roc_auc'` | Fraud detection gibi, senin notlarındaki senaryo |
| Classification (dengesiz, precision/recall önemli) | `'f1'` | Pozitif sınıfı yakalamak kritikse (hastalık teşhisi gibi) |
| Regression | `'neg_mean_squared_error'` veya `'r2'` | Sayısal tahmin |
| Multiclass (3+ sınıf) | `'f1_macro'` veya `'f1_weighted'` | Sınıf sayısı 2'den fazlaysa |

**Bu kadar.** Bunları bilip pratikte kullanmaya başlaman yeterli — geri kalanı (`neg_log_loss`, `average_precision`, `adjusted_rand_score` vs.) ihtiyaç anında bakacağın şeyler, şimdiden ezberlemene gerek yok.

## Şimdilik gerekmeyen (kafanı yorma)

- Clustering metrikleri (`adjusted_rand_score` vs.) — sen henüz unsupervised learning'e girmedin
- `_micro` varyantı — `_macro`/`_weighted` çoğu zaman yeterli
- Custom scorer yazma (`make_scorer`) — ileri seviye, gerektiğinde öğrenilir

## Pratik kural (metrik seçimi karmaşasını çözer)

Aslında bu tablo bile fazla — tek bir soru sor kendine:

> **"Yanlış pozitif mi daha maliyetli, yanlış negatif mi, yoksa ikisi de eşit mi önemli?"**

- İkisi eşit önemli, sınıflar dengeli → `accuracy`
- Sınıflar dengesiz ama genel ayırt etme gücünü ölçmek istiyorum → `roc_auc`
- Yanlış negatif çok maliyetli (kanser kaçırmak gibi) → `recall`'a ağırlık veren bir metrik (`f1` veya doğrudan `recall`)
